src/System.Net.Http/src/System/Net/Http/Unix/CurlCallbacks.cs

```diff
@@ -0,0 +1,240 @@
+        }
+
+        private int CurlSendIoCtlCallback(
+            IntPtr handle,
+            int cmd,
+            IntPtr context)
+        {
+            if (cmd != curliocmd.CURLIOCMD_RESTARTREAD)
+            {
+                return curlioerr.CURLIOE_UNKNOWNCMD;
+            }
+
+            GCHandle gch = GCHandle.FromIntPtr(context);
+            var state = (RequestCompletionSource)gch.Target;
+
+            try
+            {
+                CheckDisposed();
+                state.CancellationToken.ThrowIfCancellationRequested();
+
+                Stream contentStream = state.RequestContentStream;
+                if (!contentStream.CanSeek)
+                {
+                    return curlioerr.CURLIOE_FAILRESTART;
+                }
+                contentStream.Seek(0, SeekOrigin.Begin);
+                return curlioerr.CURLIOE_OK;
+            }
+            catch (Exception ex)
+            {
+                state.SavedException = ex;
+#if DEBUG
+                if (Logging.On) Logging.Exception(Logging.Http, this, "CurlSendIoCtlCallback exception for " + Logging.GetObjectLogHash(state.ResponseMessage.ContentStream) + ": ", ex);
+#endif
+                return curlioerr.CURLIOE_FAILRESTART; // Force curl_easy_perform to fail
+            }
+        }
+
+        private ulong ExecuteCallback(
+            IntPtr buffer,
+            ulong size,
+            ulong nitems,
+            IntPtr context,
+            Func<IntPtr, ulong, RequestCompletionSource, ulong> callback)
+        {
+            size *= nitems;
+            GCHandle gch = GCHandle.FromIntPtr(context);
+            var state = (RequestCompletionSource)gch.Target;
+
+            try
+            {
+                CheckDisposed();
+#if DEBUG
+                if (Logging.On)
+                {
+                    Logging.PrintInfo(Logging.Http, this, "Executecallback for " + Logging.GetObjectLogHash(state.ResponseMessage.ContentStream) + " with " + ((size > 0) ? (size + " bytes") : " no data"));
+                }
+#endif
+                return (size == 0) ? 0 : callback(buffer, size, state);
+            }
+            catch (Exception ex)
+            {
+                state.SavedException = ex;
+#if DEBUG
+                if (Logging.On) Logging.Exception(Logging.Http, this, "Executecallback exception for " + Logging.GetObjectLogHash(state.ResponseMessage.ContentStream) + ": ", ex);
+#endif
+                return (size > 0) ? size - 1 : 1; // Force curl_easy_perform to fail
+            }
+        }
+
+        private bool TryParseStatusLine(HttpResponseMessage response, string responseHeader)
+        {
+
+            if (!responseHeader.StartsWith("HTTP/", StringComparison.OrdinalIgnoreCase))
+            {
+                return false;
+            }
+            var tokens = responseHeader.Split(new char[] { ' ' });
+            if (tokens.Length > 1)
+            {
+                response.StatusCode = (HttpStatusCode)int.Parse(tokens[1]);
```

Comment:
@stephentoub, I tried to go thru the code of Wininet HttpQueryInfoA but was not able to figure out things. Can I leave this as a TODO for now?